File: midterm/pokemon_wrapup/PokemonGameWrapUp.java
package pokemon_wrapup;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.InputMismatchException;
import java.util.Map;
import java.util.Scanner;

/**
 * The PokemonGameWrapUp class allows the user to play against a computer apponent in a turn based
 * pokemon card game
 */
public class PokemonGameWrapUp {
    final private ArrayList<Card> deck;
    final private ArrayList<Card> hand;
    final private ArrayList<Card> prizes;
    private Pokemon activePokemon;
    private Pokemon opponentActivePokemon;
    private int playerPrizesTaken;
    private int opponentPrizesTaken;
    final private boolean opponentHasBenchPokemon;
    private int opponentActiveHealth;
    final private ArrayList<Card> opponentHand;
    final private Scanner scanner;
    private boolean energyAttachedThisTurn;
    final private boolean isPlayerTurn;
    private final Map<String, Runnable> trainerCardActions;

    /**
     * Creates a new PokemonGameWrapUp game
     */
    public PokemonGameWrapUp() {
        deck = new ArrayList<>();
        hand = new ArrayList<>();
        prizes = new ArrayList<>();
        opponentHand = new ArrayList<>();
        opponentHasBenchPokemon = true;
        scanner = new Scanner(System.in);
        this.isPlayerTurn = true;

        trainerCardActions = new HashMap<>();
        trainerCardActions.put("Draw 2 Cards", this::drawTwoCards);
        trainerCardActions.put("Heal 20 Health", () -> healPokemon(activePokemon, 20));
        trainerCardActions.put("Switch Pokemon", this::switchActivePokemon);
    }

    /**
     * Fills the deck with a mix of pokemon, Energy, and Trainer cards. Then the method shuffles the deck
     */
    public void fillDeck() {
        deck.clear();

        for (int i = 0; i < 6; i++) {
            deck.add(new Charmander());
            deck.add(new Pikachu());
        }

        for (int i = 0; i < 5; i++) {
            deck.add(new Bulbasaur());
        }

        for (int i = 0; i < 5; i++) {
            deck.add(new FireEnergy());
            deck.add(new WaterEnergy());
            deck.add(new GrassEnergy());
            deck.add(new ElectricEnergy());
            deck.add(new PsychicEnergy());
        }
        
        for (int i = 0; i < 6; i++) {
            deck.add(new Trainer("Draw 2 Cards"));
            deck.add(new Trainer("Heal 20 Health"));
            deck.add(new Trainer("Switch Pokemon"));
        }

        Collections.shuffle(deck);
    }

    /**
     * Both players draw 7 cards
     * The game stops if either player doesnt draw a full hand
     */
    public void drawHand() {
        hand.clear();
        opponentHand.clear();
        
        if (!drawCards(7, hand) || !drawCards(7, opponentHand)) {
            System.exit(0);
        }

        if (!hasActivePokemon(hand)) {
            System.out.println("\nYou have no pokemon in your hand. Please start over.");
            System.exit(0);
        }

        if (!hasActivePokemon(opponentHand)) {
            System.out.println("\nOpponent has no pokemon in their hand. Please start over.");
            System.exit(0);
        }
    }

    /**
     * Draws cards from the deck and puts them in the users hand. If the player or the opponents deck is empty than a 
     * message is displayed that they lost
     *
     * @param numberOfCards the number of cards to draw from the deck
     * @param destinationHand either the usr or the opponents hand
     * @return true if the cards were successfully dealt, false if the deck was empty
     */
    private boolean drawCards(int numberOfCards, ArrayList<Card> destinationHand) {
        for (int i = 0; i < numberOfCards; i++) {
            if (deck.isEmpty()) {
                String message;
                if (destinationHand == hand) {
                    message = "\nYour deck is empty! You lose!\n";
                } else {
                    message = "\nOpponent's deck is empty! Opponent loses!\n";
                }
                System.out.println(message);
                return false;
            }
            destinationHand.add(deck.remove(0));
        }
        return true;
    }

    /**
     * Gets the active player
     * 
     * @return the active turn, opponent or user
     */
    public String getActivePlayer() {
        return isPlayerTurn ? "player" : "opponent";
    }

    /**
     * Makes sure that the hand has at least one pokemon
     * 
     * @param hand the hand being checked for a pokemon
     * @return true if the hand contains at least one pokemon and false if not
     */
   private boolean hasActivePokemon(ArrayList<Card> hand) {
    for (Card card : hand) {
        if (card instanceof Pokemon) {
            return true;
        }
    }
    return false;
}

    /**
     * Draws a card from the user’s deck and ends the game if the deck is empty.
     */
    public void drawCard() {
        if (drawCards(1, hand)) {
            System.out.println("\nYou drew a card.\n");
            System.out.println("Hand: " + hand + "\n");
        }
    }

    /**
     * Draws two cards from the deck when the trainer card is selected
     */
    public void drawTwoCards() {
        drawCards(2, hand);
        System.out.println("\nYou drew 2 cards.");
        System.out.println("Your hand is now: " + hand);
    }

    /**
     * Heals the specified pokemon
     */
    private void healPokemon(Pokemon target, int amount) {
        if (target != null) {
            target.heal(amount);
            System.out.println("\nHealed " + target.getName() + " by " + amount + " Health.");
            System.out.println("Current Health: " + target.getHealth() + "\n");
        } else {
            System.out.println("\nNo active pokemon to heal.\n");
        }
    }

    /**
     * Switches the user's active pokemon with a selected pokemon from the hand.
     */
    public void switchActivePokemon() {
        System.out.println("\nChoose a pokemon to switch to:");
        ArrayList<Pokemon> pokemonInHand = getPokemonsInHand(hand);
        if (pokemonInHand.isEmpty()) {
            System.out.println("\nYou have no pokemon in your hand to switch to.\n");
            return;
        }
        
        int choice = scanner.nextInt() - 1;
        if (choice >= 0 && choice < pokemonInHand.size()) {
            Pokemon newActivePokemon = pokemonInHand.get(choice);
            hand.remove(newActivePokemon);
            hand.add(activePokemon);
            activePokemon = newActivePokemon;
            System.out.println("You switched to: " + activePokemon.getName() + "\n");
        } else {
            System.out.println("\nInvalid choice. Please select a valid pokemon.\n");
        }
    }

    /**
     * Helper method to get pokemon in the specified hand
     * 
     * @param hand search for pokemon cards in this hand
     * @return all of the pokemon in the hand
     */
    private ArrayList<Pokemon> getPokemonsInHand(ArrayList<Card> hand) {
        ArrayList<Pokemon> pokemonInHand = new ArrayList<>();
        for (Card card : hand) {
            if (card instanceof Pokemon) {
                pokemonInHand.add((Pokemon) card);
                System.out.println((pokemonInHand.size()) + ": " + card.getName());
            }
        }
        return pokemonInHand;
    }

    /**
     * Uses a Trainer card from the player's hand
     */
    public void useTrainerCard() {
        System.out.println("Choose a Trainer card to use:");
        ArrayList<Trainer> trainerCards = new ArrayList<>();
        for (Card card : hand) {
            if (card instanceof Trainer) {
                trainerCards.add((Trainer) card);
                System.out.println((trainerCards.size()) + ": " + card.getName());
            }
        }

        if (trainerCards.isEmpty()) {
            System.out.println("You have no Trainer cards in your hand.");
            return;
        }

        int choice = scanner.nextInt() - 1;
        if (choice >= 0 && choice < trainerCards.size()) {
            Trainer selectedTrainer = trainerCards.get(choice);
            hand.remove(selectedTrainer);
            trainerCardActions.getOrDefault(selectedTrainer.getName(), () -> System.out.println("\nTrainer card not recognized.\n")).run();
        } else {
            System.out.println("Invalid choice. Please select a valid Trainer card.");
        }
    }

    /**
     * Attaches an energy card to the chosen pokemon in the hand
     * 
     * @param target the pokemon that the energy will be attached to
     * @param hand the hand the energy card is drawn from
     */
    private void attachEnergy(Pokemon target, ArrayList<Card> hand) {
        if (target == null) {
            System.out.println("\nNo active pokemon to attach energy.\n");
            return;
        }

        Card energyCard = hand.stream().filter(card -> card instanceof Energy).findFirst().orElse(null);
        if (energyCard != null) {
            hand.remove(energyCard);
            target.attachEnergy((Energy) energyCard);
        } else {
            System.out.println("\nNo Energy card available in hand.\n");
        }
    }

    /**
     * Attaches an energy card to the opponents pokemon
     */
    public void opponentAttachEnergy() {
        if (opponentActivePokemon == null) {
            System.out.println("\nOpponent must have an active pokemon to attach energy.\n");
            return;
        }

        Card energyCard = opponentHand.stream()
            .filter(card -> card instanceof Energy)
            .findFirst()
            .orElse(null);

        if (energyCard != null) {
            opponentHand.remove(energyCard);
            opponentActivePokemon.attachEnergy((Energy) energyCard);
        } else {
            System.out.println("\nOpponent has no Energy card to attach.\n");
        }
    }

    /**
     * Allows the user to attack the opponents pokemon. The opponents pokemons health goes down by the amount of damage taken from attack.
     * The user claims a prize if the opopnents pokemon is knocked out
     */
    public void attack() {
        if (activePokemon.hasSufficientEnergy()) {
            System.out.println("\nAttacking the opponent...");

            int damageDealt = activePokemon.getAttackDamage();
            opponentActiveHealth -= damageDealt;
    
            if (opponentActiveHealth < 0) {
                opponentActiveHealth = 0;
            }
    
            System.out.println("You dealt " + damageDealt + " damage to opponent's " + opponentActivePokemon + ". Opponent's Health: " + opponentActiveHealth);
    
            if (opponentActiveHealth <= 0) {
                claimPrize();
                if (playerPrizesTaken==6) {
                    return;
                }
                replaceOpponentActivePokemon();
            }
        } else {
            System.out.println("Not enough energy to attack.");
        }
    }
    
    /**
     * Gives the user a prize card and adds it to the player's hand.
     */
    public void claimPrize() {
        if (!prizes.isEmpty()) {
            Card prizeCard = prizes.remove(0);
            hand.add(prizeCard);
            playerPrizesTaken++;
            System.out.println("You took a prize card: " + prizeCard.getName());
        }
    }

    /**
     * Gives the opponent a prize card if they knock out the users pokemon
     */
    public void claimOpponentPrize() {
        if (!deck.isEmpty()) {
            Card prizeCard = deck.remove(0);
            opponentHand.add(prizeCard);
            opponentPrizesTaken++;
            System.out.println("Opponent took a prize card.");
        }
    }

    /**
     * Checks to see if anyone has won the game yet. You win by taking all 6 prize cards.
     *
     * @return true if the game is over, false if it is not
     */
    public boolean checkGameState() {
        if (playerPrizesTaken == 6) {
            System.out.println("\nCongratulations! You won by taking all 6 of your prize cards!\n");
            return true;
        }
        
        if (opponentPrizesTaken == 6) {
            System.out.println("\nYou lost! Your opponent took all 6 prize cards.\n");
            return true;
        }
        
        if (!opponentHasBenchPokemon) {
            System.out.println("\nCongratulations! You won! Your opponent has no benched pokemon left.\n");
            return true;
        }
        
        return false;
    }

    /**
     * This method starts the game, going back and forth between the user and opponents turns checking to see if anyone
     * has won yet
     */
    public void startGame() {
        setupGame();
        setActivePokemon();
        while (true) {
            performPlayerTurn();
            if (checkGameState()) break;
            performOpponentTurn();
            if (checkGameState()) break;
        }
    }

    /**
     * Chooses the players active pokemon(first pokemon in their hand)
     * If there are no pokemon in the hand it displays a message saying that there is no pokemon to play.
     */
    private void setActivePokemon() {
        for (Card card : hand) {
            if (card instanceof Pokemon) {
                activePokemon = (Pokemon) card;
                System.out.println("\nYour Hand: " + hand);
                System.out.println("Your active pokemon is: " + activePokemon.getName());
                hand.remove(card);
                return;
            }
        }
        System.out.println("You do not have any pokemon in your hand to play!");
    }

    /**
     * Gets the game ready by filling and shuffling the deck, drawing hands, and setting active pokemon
     * Draws 6 cards as prize cards
     */
    public void setupGame() {
        fillDeck();
        drawHand();
        activePokemon = (Pokemon) hand.stream().filter(card -> card instanceof Pokemon).findFirst().orElse(null);
        opponentActivePokemon = (Pokemon) opponentHand.stream().filter(card -> card instanceof Pokemon).findFirst().orElse(null);
        if (activePokemon != null) {
            activePokemon.setActive(true);
        }
        if (opponentActivePokemon != null) {
            opponentActivePokemon.setActive(true);
            opponentActiveHealth = opponentActivePokemon.getHealth();
        }
        for (int i = 0; i < 6; i++) {
            prizes.add(deck.remove(0));
        }
    }

    private boolean isFirstTurn = true;

    /**
     * Performs the player's turn, lets the user attach energy, attack, use a Trainer card, or end their turn.
     */
    public void performPlayerTurn() {
    System.out.println("\n--- Your Turn ---");

    if (!isFirstTurn) {
        drawCard();
    }

    isFirstTurn = false;
    energyAttachedThisTurn = false;

    if (activePokemon != null) {
        System.out.println("Energy attached to " + activePokemon.getName() + ": " + activePokemon.getTotalEnergy());
    }

    while (true) {
        try {
            System.out.println("Choose an action: (1) Attach Energy, (2) Attack, (3) Use Trainer Card, (4) End Turn");
            int choice = scanner.nextInt(); 

            switch (choice) {
                case 1:
                    if (!energyAttachedThisTurn) {
                        attachEnergy(activePokemon, hand);  // Pass activePokemon and hand as arguments
                        energyAttachedThisTurn = true;
                    } else {
                        System.out.println("\nYou can only attach one Energy card per turn.");
                    }
                    break;
                case 2:
                    attack();
                    return;
                case 3:
                    if (hasTrainerCardInHand()) {
                        useTrainerCard();
                    } else {
                        System.out.println("You have no Trainer cards in your hand. Please choose another action.");
                    }
                    return;
                case 4:
                    System.out.println("Ending your turn.");
                    return;
                default:
                    System.out.println("Invalid choice. Please choose a valid action.");
            }
        } catch (InputMismatchException e) {
            System.out.println("Invalid input! Please enter a number");
            scanner.nextLine();
        }
    }
}


    /**
     * Looks for a trainer card in the users hand
     *
     * @return true if there is a Trainer card in the hand, false otherwise
     */
    private boolean hasTrainerCardInHand() {
        return hand.stream().anyMatch(card -> card instanceof Trainer);
    }

    /**
     * Performs the opponents turn by drawing a card, attaching energy, and attacking
     */
    public void performOpponentTurn() {
        if (!isFirstTurn) {
            opponentDrawCard();
        }

        isFirstTurn = false;
        System.out.println("\n--- Opponent's Turn ---");
        System.out.println("Opponent drew a card.");

        // Opponent attaches energy before attacking
        opponentAttachEnergy();

        if (opponentActivePokemon != null) {
            System.out.println("Energy attached to " + opponentActivePokemon.getName() + ": " + opponentActivePokemon.getTotalEnergy());
        }

        // Check if opponent's pokemon has enough energy to attack
        if (opponentActivePokemon.hasSufficientEnergy()) {
            System.out.println("Opponent's " + opponentActivePokemon.getName() + " attacks!");

            if (activePokemon != null) {
                activePokemon.takeDamage(opponentActivePokemon.getAttackDamage());
                System.out.println("Your pokemon now has " + activePokemon.getHealth() + " Health remaining.");
                if (activePokemon.getHealth() <= 0) {
                    System.out.println("Your pokemon is knocked out!");
                    replaceActivePokemon();
                    claimOpponentPrize();

                    if (checkGameState()) return;
                }
            }
        } else {
            System.out.println("Opponent's " + opponentActivePokemon.getName() + " does not have enough energy to attack.");
        }
    }
    
    /**
     * Draws a card from the opponents deck
     * If the deck is empty the opponent loses
     */
    private void opponentDrawCard() {
        if (!deck.isEmpty()) {
            Card drawnCard = deck.remove(0);
            opponentHand.add(drawnCard);
        } else {
            System.out.println("\nOpponent's deck is empty! Opponent loses!");
            System.exit(0);
        }
    }

    /**
     * Replaces the opponent's active pokemon with another pokemon from their hand, if available. If no replacement 
     * is available a message is shown that the opponent has no pokemon left to play.
     */
    private void replaceOpponentActivePokemon() {
        // Clear the energy from the knocked-out pokemon
        if (opponentActivePokemon != null) {
            opponentActivePokemon.clearEnergy(); // Reset energy when pokemon is knocked out
        }
    
        opponentHand.add(opponentActivePokemon);
        opponentActivePokemon = null;
    
        // Find a new active pokemon from the opponent's hand
        for (Card card : opponentHand) {
            if (card instanceof Pokemon) {
                opponentActivePokemon = (Pokemon) card;
                System.out.println("\nOpponent's pokemon knocked out!");
                System.out.println("Opponent's new active pokemon is: " + opponentActivePokemon.getName());
                opponentHand.remove(card);
                opponentActiveHealth = opponentActivePokemon.getHealth();
                opponentActivePokemon.setActive(true);
                return;
            }
        }
        System.out.println("Opponent does not have any pokemon left to play!");
    }
    
    /**
     * Draws a card from the opponents deck. Ends the game if the deck is empty
     */
    public void drawAdditionalOpponentCards(int numberOfCards) {
        for (int i = 0; i < numberOfCards; i++) {
            if (!deck.isEmpty()) {
                opponentHand.add(deck.remove(0));
            } else {
                System.out.println("\nOpponent's deck is empty! Cannot draw more cards.");
                break;
            }
        }
    }

    /**
     * Heals the opponents pokemon
     *
     * @param amount the amount of health given to the opponents pokemon
     */
    public void healOpponentActivePokemon(int amount) {
        if (opponentActivePokemon != null) {
            opponentActivePokemon.heal(amount);
            System.out.println("Opponent healed " + opponentActivePokemon.getName() + " by " + amount + " Health.");
            System.out.println("Opponent's pokemon Health: " + opponentActivePokemon.getHealth());
        } else {
            System.out.println("Opponent has no active pokemon to heal.");
        }
    }

    /**
     * Changes the users active pokemon to a different pokemon from their deck. If the active pokemon is 
     * knocked out a new pokemon is set as active.
     */
    public void replaceActivePokemon() {
        if (activePokemon != null) {
            activePokemon.clearEnergy();
        }
        hand.add(activePokemon);
        activePokemon = null;

        // look for a new pokemon in the player's hand to maake the new active pokemon
        for (Card card : hand) {
            if (card instanceof Pokemon) {
                activePokemon = (Pokemon) card;
                System.out.println("Your new active pokemon is: " + activePokemon.getName());
                hand.remove(card);
                // set new active polemon
                activePokemon.setActive(true);
                return;
            }
        }
        System.out.println("You do not have any pokemon left to play!");
    }

    /**
     * The main method that runs the game
     *
     * @param args
     */
    public static void main(String[] args) {
        PokemonGameWrapUp game = new PokemonGameWrapUp();
        game.startGame();
    }
}

File: final/part3/PokemonGameWrapUp.swift
import Foundation

class PokemonGameWrapUp: ObservableObject {
    @Published var deck: [Card] = []
    @Published var hand: [Card] = []
    @Published var opponentHand: [Card] = []
    @Published var prizes: [Card] = []
    @Published var activePokemon: Pokemon?
    @Published var opponentActivePokemon: Pokemon?
    @Published var playerPrizesTaken = 0
    @Published var opponentPrizesTaken = 0
    @Published var opponentHasBenchPokemon = true
    @Published var isPlayerTurn = true
    @Published var gameOverMessage: String?
    @Published var lastAttackMessage: String = ""

    private var opponentActiveHealth = 0
    private var energyAttachedThisTurn = false
    private var isFirstTurn = true
    private var trainerCardActions: [String: () -> Void] = [:]

    init() {
        setupTrainerCardActions()
        setupGame()
    }

    // MARK: - Setup Methods

    private func setupTrainerCardActions() {
        trainerCardActions["Draw 2 Cards"] = { self.drawTwoCards() }
        trainerCardActions["Heal 20 Health"] = { self.healPokemon(self.activePokemon, amount: 20) }
        trainerCardActions["Switch Pokemon"] = { self.switchActivePokemon() }
    }

    func fillDeck() {
        deck.removeAll()

        for _ in 0..<6 {
            deck.append(Charmander())
            deck.append(Pikachu())
        }

        for _ in 0..<5 {
            deck.append(Bulbasaur())
        }

        for _ in 0..<5 {
            deck.append(FireEnergy())
            deck.append(WaterEnergy())
            deck.append(GrassEnergy())
            deck.append(ElectricEnergy())
            deck.append(PsychicEnergy())
        }

        for _ in 0..<6 {
            deck.append(Trainer(name: "Draw 2 Cards", action: "drawTwoCards"))
            deck.append(Trainer(name: "Heal 20 Health", action: "healPokemon"))
            deck.append(Trainer(name: "Switch Pokemon", action: "switchPokemon"))
        }

        deck.shuffle()
    }

    func setupGame() {
        fillDeck()
        drawHand()

        activePokemon = hand.first { $0 is Pokemon } as? Pokemon
        opponentActivePokemon = opponentHand.first { $0 is Pokemon } as? Pokemon

        activePokemon?.isActive = true
        opponentActivePokemon?.isActive = true
        opponentActiveHealth = opponentActivePokemon?.health ?? 0

        for _ in 0..<6 {
            prizes.append(deck.removeFirst())
        }
    }

    private func drawHand() {
        hand.removeAll()
        opponentHand.removeAll()

        guard drawCards(count: 7, destination: &hand),
              drawCards(count: 7, destination: &opponentHand) else {
            print("Failed to draw hands. Game over.")
            exit(0)
        }

        if !hasActivePokemon(hand) {
            print("You have no Pokémon in your hand. Please start over.")
            exit(0)
        }

        if !hasActivePokemon(opponentHand) {
            print("Opponent has no Pokémon in their hand. Please start over.")
            exit(0)
        }
    }

    func drawCards(count: Int, destination: inout [Card]) -> Bool {
        for _ in 0..<count {
            guard !deck.isEmpty else {
                let message = (destination as AnyObject === hand as AnyObject)
                    ? "Your deck is empty! You lose!"
                    : "Opponent's deck is empty! You win!"
                print(message)
                return false
            }
            destination.append(deck.removeFirst())
        }
        return true
    }


    private func hasActivePokemon(_ hand: [Card]) -> Bool {
        hand.contains { $0 is Pokemon }
    }

    // MARK: - Player Actions

    func drawCard() {
        if drawCards(count: 1, destination: &hand) {
            print("You drew a card. Hand: \(hand.map { $0.name })")
        }
    }

    func drawTwoCards() {
        if drawCards(count: 2, destination: &hand) {
            print("You drew 2 cards. Hand: \(hand.map { $0.name })")
        }
    }

    func healPokemon(_ target: Pokemon?, amount: Int) {
        guard let target = target else {
            print("No active Pokémon to heal.")
            return
        }
        target.heal(amount)
        print("Healed \(target.name) by \(amount). Current health: \(target.health).")
    }

    func attachEnergy(to target: Pokemon?) {
        guard let target = target else {
            print("No active Pokémon to attach energy.")
            return
        }
        guard let energyCard = hand.first(where: { $0 is Energy }) as? Energy else {
            print("No energy cards in hand.")
            return
        }
        hand.removeAll { $0.name == energyCard.name }
        target.attachEnergy(energyCard)
        print("Attached \(energyCard.name) to \(target.name).")
    }

    func switchActivePokemon() {
        let pokemonInHand = hand.compactMap { $0 as? Pokemon }
        guard !pokemonInHand.isEmpty else {
            print("No Pokémon in hand to switch to.")
            return
        }

        for (index, pokemon) in pokemonInHand.enumerated() {
            print("\(index + 1): \(pokemon.name)")
        }

        if let choice = readLine(), let selectedIndex = Int(choice), selectedIndex > 0, selectedIndex <= pokemonInHand.count {
            let selectedPokemon = pokemonInHand[selectedIndex - 1]
            hand.removeAll { $0.name == selectedPokemon.name }
            if let active = activePokemon {
                hand.append(active)
            }
            activePokemon = selectedPokemon
            print("Switched to \(selectedPokemon.name).")
        } else {
            print("Invalid choice.")
        }
    }

    func attack() {
        guard let attacker = activePokemon, let defender = opponentActivePokemon else {
            lastAttackMessage = "No Pokémon to attack."
            return
        }

        if attacker.hasSufficientEnergy() {
            let damage = attacker.attackDamage
            defender.takeDamage(damage)
            lastAttackMessage = "\(attacker.name) dealt \(damage) damage to \(defender.name). Remaining health: \(defender.health)"

            if defender.health <= 0 {
                lastAttackMessage += "\n\(defender.name) has been knocked out!"
                claimPrize()
                replaceOpponentActivePokemon()
            }
        } else {
            lastAttackMessage = "\(attacker.name) does not have enough energy to attack."
        }
    }





    func useTrainerCard() {
        let trainersInHand = hand.compactMap { $0 as? Trainer }
        guard !trainersInHand.isEmpty else {
            print("No trainer cards in hand.")
            return
        }

        for (index, trainer) in trainersInHand.enumerated() {
            print("\(index + 1): \(trainer.name)")
        }

        if let choice = readLine(), let selectedIndex = Int(choice), selectedIndex > 0, selectedIndex <= trainersInHand.count {
            let selectedTrainer = trainersInHand[selectedIndex - 1]
            hand.removeAll { $0.name == selectedTrainer.name }
            trainerCardActions[selectedTrainer.name]?()
        } else {
            print("Invalid choice.")
        }
    }

    // MARK: - Opponent Actions

    func opponentAttachEnergy() {
        guard let target = opponentActivePokemon else {
            print("Opponent has no active Pokémon to attach energy.")
            return
        }

        if let energyCard = opponentHand.first(where: { $0 is Energy }) as? Energy {
            opponentHand.removeAll { $0.name == energyCard.name }
            target.attachEnergy(energyCard)
            print("Opponent attached \(energyCard.name) to \(target.name).")
        } else {
            print("Opponent has no energy cards.")
        }
    }

    func performOpponentTurn() {
        print("\n--- Opponent's Turn ---")
        if drawCards(count: 1, destination: &opponentHand) {
            print("Opponent drew a card.")
        }
        opponentAttachEnergy()

        if let opponentPokemon = opponentActivePokemon, opponentPokemon.hasSufficientEnergy() {
            attackOpponent()
        } else {
            print("Opponent cannot attack.")
        }
    }

    private func attackOpponent() {
        guard let attacker = opponentActivePokemon, let defender = activePokemon else { return }
        let damage = attacker.attackDamage
        defender.takeDamage(damage)
        print("Opponent's \(attacker.name) attacked! \(defender.name) has \(defender.health) health remaining.")

        if defender.health <= 0 {
            print("\(defender.name) has been knocked out!")
            claimOpponentPrize()
            replaceActivePokemon()
        }
    }

    func claimPrize() {
        guard !prizes.isEmpty else { return }
        let prize = prizes.removeFirst()
        hand.append(prize)
        playerPrizesTaken += 1
        print("Claimed prize card: \(prize.name).")
    }

    func claimOpponentPrize() {
        opponentPrizesTaken += 1
        print("Opponent claimed a prize card.")
    }

    func replaceOpponentActivePokemon() {
        guard let knockedOutPokemon = opponentActivePokemon else { return }

        // Reset the knocked-out Pokémon's energy
        knockedOutPokemon.clearEnergy()

        // Add the knocked-out Pokémon back to the opponent's hand
        opponentHand.append(knockedOutPokemon)

        // Find a new active Pokémon
        if let newActivePokemon = opponentHand.first(where: { $0 is Pokemon }) as? Pokemon {
            opponentActivePokemon = newActivePokemon
            opponentHand.removeAll { $0.name == newActivePokemon.name }
            print("Opponent's new active Pokémon is \(newActivePokemon.name).")
        } else {
            // If no Pokémon are available, end the opponent's ability to fight
            opponentActivePokemon = nil
            opponentHasBenchPokemon = false
            print("Opponent has no Pokémon left to play!")
        }
    }


    private func replaceActivePokemon() {
        activePokemon = hand.compactMap { $0 as? Pokemon }.first
        if let newActive = activePokemon {
            hand.removeAll { $0.name == newActive.name }
            print("Switched to \(newActive.name).")
        } else {
            print("No Pokémon left to play!")
        }
    }

    // MARK: - Game State

    func checkGameState() -> Bool {
        if playerPrizesTaken == 6 {
            print("You win! All prizes claimed.")
            return true
        }

        if opponentPrizesTaken == 6 || !opponentHasBenchPokemon {
            print("You lose! Opponent won.")
            return true
        }

        return false
    }

    func startGame() {
        setupGame()

        while true {
            performPlayerTurn()
            if checkGameState() { break }
            performOpponentTurn()
            if checkGameState() { break }
        }
    }

    func performPlayerTurn() {
        print("\n--- Your Turn ---")

        if !isFirstTurn {
            drawCard()
        }

        isFirstTurn = false
        energyAttachedThisTurn = false

        if let active = activePokemon {
            print("Your active Pokémon is \(active.name).")
            print("Energy attached: \(active.totalEnergy)")
            print("Health: \(active.health)")
        } else {
            print("You have no active Pokémon.")
            return
        }

        while true {
            print("\nChoose an action:")
            print("(1) Attach Energy")
            print("(2) Attack")
            print("(3) Use Trainer Card")
            print("(4) End Turn")
            print("Your hand: \(hand.map { $0.name })")

            if let choice = Int(readLine() ?? "0") {
                switch choice {
                case 1:
                    if !energyAttachedThisTurn {
                        attachEnergy(to: activePokemon)
                        energyAttachedThisTurn = true
                    } else {
                        print("You can only attach one Energy card per turn.")
                    }
                case 2:
                    attack()
                    return
                case 3:
                    useTrainerCard()
                    return
                case 4:
                    print("Ending your turn.")
                    isPlayerTurn.toggle()
                    return
                default:
                    print("Invalid choice. Please choose a valid action.")
                }
            } else {
                print("Invalid input. Please enter a number.")
            }
        }
    }
}
